src/main/java/com/es/phoneshop/model/cart/HttpSessionCartService.java

```diff
@@ -0,0 +1,77 @@
+package com.es.phoneshop.model.cart;
+
+import com.es.phoneshop.exceptions.OutOfStockException;
+import com.es.phoneshop.model.product.Product;
+
+import javax.servlet.http.HttpServletRequest;
+import java.math.BigDecimal;
+import java.util.Optional;
+
+public class HttpSessionCartService implements CartService {
+    private static final String CART_ATTRIBUTE = CartService.class + ".cart";
+
+    private HttpSessionCartService() {
+    }
+
+    public static HttpSessionCartService getInstance() {
+        return CartServiceHolder.cartService;
+    }
+
+    public Cart getCart(HttpServletRequest request) {
+        Cart cart = (Cart) request.getSession().getAttribute(CART_ATTRIBUTE);
+        if (cart == null) {
+            cart = new Cart();
+            request.getSession().setAttribute(CART_ATTRIBUTE, cart);
+        }
+
+        return cart;
+    }
+
+    public void add(Cart cart, Product product, int quantity) {
```

Comment:
Refactor this method to be more Java-8 style. Make use of Optional.map, Optional.orElse, etc. E.g., the total quantity can be represented by:
totalQuantity = quantity + optionalItem
    .map(CartItem::getQuantity)
    .orElse(0);
Maybe it's possible to refactor this whole logic into one / two simple optional chains